src/Symfony/Component/Runtime/SymfonyRuntime.php

```diff
@@ -0,0 +1,198 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Runtime;
+
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\ArgvInput;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\ConsoleOutput;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Dotenv\Dotenv;
+use Symfony\Component\ErrorHandler\Debug;
+use Symfony\Component\ErrorHandler\ErrorHandler;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\HttpKernel\TerminableInterface;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class SymfonyRuntime implements RuntimeInterface
+{
+    private $request;
+    private $input;
+    private $output;
+    private $application;
+    private $command;
+
+    public function __construct(string $projectDir = null)
+    {
+        $projectDir = $projectDir ?? \dirname(__DIR__, 3);
+
+        if (!class_exists(Dotenv::class)) {
+            throw new \LogicException(sprintf('You cannot use "%s" as the Dotenv component is not installed. Try running "composer require symfony/dotenv".', __CLASS__));
+        }
+
+        if (!class_exists(ErrorHandler::class)) {
+            throw new \LogicException(sprintf('You cannot use "%s" as the ErrorHandler component is not installed. Try running "composer require symfony/error-handler".', __CLASS__));
+        }
+
+        if (isset($_SERVER['argv']) && class_exists(ArgvInput::class)) {
+            $this->getInput();
+        }
+
+        (new Dotenv())->bootEnv($projectDir.'/.env');
+
+        if ($_SERVER['APP_DEBUG']) {
+            umask(0000);
+            Debug::enable();
+        } else {
+            ErrorHandler::register();
+        }
+    }
+
+    public function resolveMain(\Closure $main): array
+    {
+        $arguments = [];
+
+        try {
+            foreach ((new \ReflectionFunction($main))->getParameters() as $parameter) {
+                $arguments[] = $this->getArgument($parameter);
+            }
+        } catch (\InvalidArgumentException $e) {
+            if (!$parameter->isOptional()) {
+                throw $e;
+            }
+        }
+
+        return [$main, $arguments];
+    }
+
+    public function getHandler(object $result): \Closure
+    {
+        if ($result instanceof HttpKernelInterface) {
```

Comment:
This code looks like multiple runtimes glued to single one just to name it `SymfonyRuntime`. I suppose it should be a bit hard to support and extend this class in case you need to improve single runtime to your custom needs. I suggest to split it to smaller runtimes with `SymfonyRuntime`

Reply:
Specialized runtimes are not really interesting from a consumer pov. A generic runtime provides better decoupling actually, from the app pov: "no need to care about the runtime, it just works."

About extensibility, I don't see any limitations. Decoration works just great to me. Can you elaborate on which limitation you spotted?

Reply:
Currently I can't override behaviour just of `Application` handler without copy-pasting entire if clause and maintaining own copy all the way further. I.e to add some `ini_set` clauses alongside pre-defined `set_time_limit`.

> generic runtime

Looks like I failed to finish my proposal in previous message. I suggest to splite them to smaller runtimes and then aggregate back into big one using some kind of pattern. Quick draft:

```
intraface:
...
public function getHandler(object $result): ?\Closure
```
```
HttpKernelInterfaceRuntime:
if ($result instance of ...) return ...
return null;
```
```
ResponseRuntime:
if ($result instance of ...) return ...
return null
```

```
SymfonyRuntime:
gteHandler:
foreach ($this->runtimes as runtime) {
  $handler = $runtime->getHandler($result)
  if ($handler  !== null) return $handler;
} 

return null; // (or some kind of exception throwing handler)
```

Idea is that end-user can compose own runtime re-using your ones as a complete class, not hooking up into inheritance chain

Reply:
> I can't override behaviour just of Application handler without copy-pasting

Not sure what you mean. This does it:
```php
function getHandler($result)
{
	if ($result instanceof Application) {
		// do your stuff
	} else {
		// call SymfonyRuntime::getHandler() 
	}
}
```

> suggest to splite them to smaller runtimes and then aggregate back into big one using some kind of pattern.

I experimented with this idea in https://github.com/tchwork/bootstrapper#extensibility and wasn't convinced it should make it into the first iteration at least. I'm not sure it's worth the overhead. But that's not a limitation since that can be provided by a third-party package and/or be added in this one in the future if we have some proof that it's worth it.

Reply:
> This does it:

Since this moment (If I implement this) I lost any updates and patches from symfony/runtime package for this type of result and live on my own. This is sad for me (as developer). This is not an extensibility, I think

Reply:
You didn't lose anything: you can also call `SymfonyRuntime::getHandler()` in the "if" branch. There is zero difference with any other design in terms of extensibility - that's the beauty of high-order interfaces.